components/ht16k33/ht16k33.c

```diff
@@ -0,0 +1,151 @@
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+#include <memory.h>
+
+#include <freertos/FreeRTOS.h>
+
+#include <esp_check.h>
+#include <esp_err.h>
+#include <esp_log.h>
+
+#include "ht16k33.h"
+
+static const char* TAG = "ht16k33";
+
+// All commands are one byte long.
+// First 4 bits are command type.
+// Last 4 bits are parameters.
+#define HT16K33_CMD_RAM_SET_POINTER 0b0000
+#define HT16K33_CMD_SYSTEM_SETUP 0b0010
+#define HT16K33_CMD_SET_BRIGHTNESS 0b1110
+#define HT16K33_CMD_DISPLAY_SETUP 0b1000
+#define HT16K33_CMD_ROW_INT_SET 0b1010
+
+#define HT16K33_SET_RAM_CMD_SIZE_BYTES HT16K33_RAM_SIZE_BYTES + 1
+
+esp_err_t zero_ram(i2c_dev_t *dev) {
+    uint8_t all_zeros[HT16K33_RAM_SIZE_BYTES];
+    memset(all_zeros, 0, HT16K33_RAM_SIZE_BYTES);
+
+    return ht16k33_ram_write(dev, all_zeros);
+}
+
+// Send one-byte command.
+esp_err_t send_short_cmd(i2c_dev_t *dev, uint8_t cmd) {
+    I2C_DEV_TAKE_MUTEX(dev);
+    I2C_DEV_CHECK(dev, i2c_dev_write(dev, NULL, 0, &cmd, sizeof(uint8_t)));
+    I2C_DEV_GIVE_MUTEX(dev);
+
+    return ESP_OK;
+}
+
+esp_err_t ht16k33_init(i2c_dev_t *dev, i2c_port_t port, uint32_t i2c_freq_hz,
+ gpio_num_t sda_gpio, gpio_num_t scl_gpio, uint8_t addr)
+{
+    memset(dev, 0, sizeof(i2c_dev_t));
+    dev->port = port;
+    dev->addr = addr;
+    dev->cfg.sda_io_num = sda_gpio;
+    dev->cfg.scl_io_num = scl_gpio;
+    dev->cfg.master.clk_speed = i2c_freq_hz;
+
+    ESP_RETURN_ON_ERROR(i2c_dev_create_mutex(dev), TAG, "Can't initialize I2C mutex.");
+
+    // Enable system oscillator.
+    ESP_RETURN_ON_ERROR(
+        send_short_cmd(dev, HT16K33_CMD_SYSTEM_SETUP << 4 | 0b0001),
+        TAG,
+        "Can't start oscillator.");
+
+    // Set to ROW output.
+    ESP_RETURN_ON_ERROR(
+        send_short_cmd(dev, HT16K33_CMD_ROW_INT_SET << 4 | 0b0000),
+        TAG,
+        "Can't set to ROW output");
+
+    // Set half brightness.
+    ESP_RETURN_ON_ERROR(
+        ht16k33_set_brightness(dev, HT16K33_MAX_BRIGHTNESS / 2 ),
+        TAG,
+        "Can't set initial brightness.");
+
+    // Set display off, no blinking.
+    ESP_RETURN_ON_ERROR(
+        ht16k33_display_setup(dev, 0, HT16K33_BLINKING_0HZ),
+        TAG,
+        "Can't set display off, no blinking.");
+
+    // RAM initializes with random values. Zero it.
+    ESP_RETURN_ON_ERROR(zero_ram(dev), TAG, "Can't zero the RAM.");
+
+    return ESP_OK;
+}
+
+esp_err_t ht16k33_free_desc(i2c_dev_t *dev)
+{
+    return i2c_dev_delete_mutex(dev);
+}
+
+esp_err_t ht16k33_set_brightness(i2c_dev_t *dev, uint8_t brightness) {
```

Comment:
Please move open parentheses to the next line in function declarations, as in `ht16k33_free_desc()`.

Reply:
It looks like `clang-format` 14 keeps opening parenthesis at the function declaration line by default.  
What do you think? I suppose its better to keep the default setting.